### PR TITLE
8264666: Change implementation of safeAdd/safeMult in the LCMSImageLayout class

### PR DESCRIPTION
**Pull request by mrserb**
Description of the new version of the fix:
While I have worked on this change and tried to consider the comments, I have found that the usages of the "safeAdd/safeMult" in the LCMSImageLayout class are incorrect. Both methods are based on the "Math" versions but throw a different exception. The problem is that its implementation may accept the negative values during intermediate calculation, see the old implementation of "[verify](https://github.com/openjdk/jdk/blob/139615b1815d4afd3593536d83fa8b25430f35e7/src/java.desktop/share/classes/sun/java2d/cmm/lcms/LCMSImageLayout.java#L343)" method:
 1. We check the "offset" value: 0 <= offset <  dataArrayLength
 2. We do some intermediate calculations that "accept" negative values
 3. We check the final "off" value: 0 <= offset <  dataArrayLength

I wondered is it possible to provide some data that using wrong/negative data at step2 may result in the correct check at step3. I spent some time and was able to reproduce the problem with the attached test case. Note that the test is a little bit cryptic since it is not possible to reproduce it by input image data.

Note: I have removed all cleanup from the fix, to make it simpler.

<======>
Description of the old version of the fix:
- The hand-crafted methods for addition and multiplication are replaced by the "Math" versions.
- Cleanup: the usage of do/while(false) is removed

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264666](https://bugs.openjdk.java.net/browse/JDK-8264666): Change implementation of safeAdd/safeMult in the LCMSImageLayout class

### Reviewers
 * [Phil Race](https://openjdk.java.net/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3333/head:pull/3333` \
`$ git checkout pull/3333`

Update a local copy of the PR: \
`$ git checkout pull/3333` \
`$ git pull https://git.openjdk.java.net/jdk pull/3333/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3333`

View PR using the GUI difftool: \
`$ git pr show -t 3333`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3333.diff">https://git.openjdk.java.net/jdk/pull/3333.diff</a>

</details>

### CODE REVIEW

**Comment by mrserb** [comment]
The nice PR number: 3333!

File: src/java.desktop/share/classes/sun/java2d/cmm/lcms/LCMSImageLayout.java [hunk 1]
```diff
@@ -222,81 +223,67 @@ public static LCMSImageLayout createImageLayout(BufferedImage image) throws Imag
...
-                    l.nextPixelOffset = safeMult(4, intRaster.getPixelStride());
-                    l.offset = safeMult(4, intRaster.getDataOffset(0));
-                    l.dataArray = intRaster.getDataStorage();
-                    l.dataArrayLength = 4 * intRaster.getDataStorage().length;
-                    l.dataType = DT_INT;
-
-                    if (l.nextRowOffset == l.width * 4 * intRaster.getPixelStride()) {
-                        l.imageAtOnce = true;
-                    }
-                } while (false);
```
Comment (mrserb) [inline src/java.desktop/share/classes/sun/java2d/cmm/lcms/LCMSImageLayout.java:241]: Any idea why the "do/while(false)" was used here?

**Comment by prrace** [comment]
I don't understand the exception handling here. multiplyExact and addExact will throw a runtime exception - ArithmeticException. And I don't see where it is caught. The change for ImageLayoutException to extend ArithmeticException doesn't seem like it can help here. Where is the ArithmeticException being turned into ImageLayoutException ?

File: src/java.desktop/share/classes/sun/java2d/cmm/lcms/LCMSImageLayout.java (hunk 1 above)
Comment (azvegint) [reply src/java.desktop/share/classes/sun/java2d/cmm/lcms/LCMSImageLayout.java:241]: Probably it is just a way to highlight and improve readability of a block of code, but simply `{}` would be enough in this case.

**Comment by mrserb** [comment]
> I don't understand the exception handling here. multiplyExact and addExact will throw a runtime exception - ArithmeticException. And I don't see where it is caught. The change for ImageLayoutException to extend ArithmeticException doesn't seem like it can help here. Where is the ArithmeticException being turned into ImageLayoutException ?

O_o for some reason... missed that.

**Comment by mrserb** [comment]
> @mrserb This pull request has been inactive for more than 8 weeks and will be automatically closed if another 8 weeks passes without any activity. To avoid this, simply add a new comment to the pull request. Feel free to ask for assistance if you need help with progressing this pull request towards integration!

not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
not now

**Comment by mrserb** [comment]
I think the usage of these methods in this class is wrong. Both allow negative values which should not be accepted.

**Comment by mrserb** [comment]
The fix is reworked based on the new facts about that methods.

**Review by prrace — APPROVED** [review]

**Pull request closed without merging**